bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/items/ItemBuilderImpl.java

```diff
@@ -0,0 +1,154 @@
+    private List<String> tags = Collections.emptyList();
+    private @Nullable String label;
+    private @Nullable Item baseItem;
+    private @Nullable GroupFunction groupFunction;
+
+    public ItemBuilderImpl(Set<ItemFactory> itemFactories, Item item) {
+        this.itemFactories = itemFactories;
+        this.itemType = item.getType();
+        this.name = item.getName();
+        this.category = item.getCategory();
+        this.groups = item.getGroupNames();
+        this.label = item.getLabel();
+        if (item instanceof GroupItem) {
+            this.baseItem = ((GroupItem) item).getBaseItem();
+            this.groupFunction = ((GroupItem) item).getFunction();
+        }
+    }
+
+    public ItemBuilderImpl(Set<ItemFactory> itemFactories, String itemType, String itemName) {
+        this.itemFactories = itemFactories;
+        this.itemType = itemType;
+        this.name = itemName;
+    }
+
+    @Override
+    public ItemBuilder withLabel(@Nullable String label) {
+        this.label = label;
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withGroups(@Nullable Collection<String> groups) {
+        if (groups != null) {
+            this.groups = new LinkedList<>(groups);
+        } else {
+            this.groups = Collections.emptyList();
+        }
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withGroups(String... groups) {
+        this.groups = Arrays.asList(groups);
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withCategory(@Nullable String category) {
+        this.category = category;
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withBaseItem(@Nullable Item item) {
+        if (item != null && !GroupItem.TYPE.equals(itemType)) {
+            throw new IllegalArgumentException("Only group items can have a base item");
+        }
+        baseItem = item;
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withGroupFunction(@Nullable GroupFunction function) {
+        if (function != null && !GroupItem.TYPE.equals(itemType)) {
+            throw new IllegalArgumentException("Only group items can have a base item");
+        }
+        groupFunction = function;
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withTags(String... tags) {
+        return this;
+    }
+
+    @Override
+    public ItemBuilder withTags(@Nullable Collection<String> tags) {
+        if (tags != null) {
+            this.tags = new LinkedList<>(tags);
+        } else {
+            this.tags = Collections.emptyList();
+        }
+        return this;
+    }
+
+    @Override
+    public Item build() {
+        Item item = null;
+        if (GroupItem.TYPE.equals(itemType)) {
+            item = new GroupItem(name, baseItem, groupFunction);
+        } else {
+            for (ItemFactory itemFactory : itemFactories) {
+                item = itemFactory.createItem(itemType, name);
```

Comment:
So the last one wins? This is  a pretty non-deterministic behaviour, isn't it?

Reply:
I assume if item is non null we would like to break the loop, so the first factory that can handle the item type wins.

Reply:
yes, of course 😩